### PR TITLE
Job failure policy controller support

### PR DESCRIPTION
**Pull request by clamoriniere1A**
**What this PR does / why we need it**:
Start implementing the support of the "Backoff policy and failed pod limit" in the ```JobController```  defined in https://github.com/kubernetes/community/pull/583.
This PR depends on a previous PR #48075  that updates the K8s API types.

TODO: 
* [X] Implement ```JobSpec.BackoffLimit``` support
* [x] Rebase when #48075 has been merged.
* [X] Implement end2end tests

implements https://github.com/kubernetes/community/pull/583

**Special notes for your reviewer**:

**Release note**:
```release-note
Add backoff policy and failed pod limit for a job
```

### CODE REVIEW

File: pkg/controller/job/job_controller.go
Comment (soltysh) [inline pkg/controller/job/job_controller.go]: Should be `EventTypeWarning` not `EventTypeNormal`.

File: pkg/controller/job/job_controller.go
Comment (soltysh) [inline pkg/controller/job/job_controller.go]: Should be `EventTypeWarning` not `EventTypeNormal`.

Comment (soltysh) [reply pkg/controller/job/job_controller.go]: Don't create methods like this, just record event and save condition in tmp variable.

File: pkg/controller/job/job_controller.go
Comment (soltysh) [reply pkg/controller/job/job_controller.go]: Don't create methods like this, just record event and save condition in tmp variable.

File: pkg/controller/job/job_controller_test.go
Comment (soltysh) [inline pkg/controller/job/job_controller_test.go]: This is generally not possible situation, since if everything goes through API it will get defaulted. So I'd just ignore this case.

File: pkg/controller/job/job_controller_test.go
Comment (soltysh) [inline pkg/controller/job/job_controller_test.go]: Same here.

File: pkg/controller/job/job_controller_test.go
Comment (soltysh) [inline pkg/controller/job/job_controller_test.go]: Not needed, drop this.

File: pkg/controller/job/job_controller.go
Comment (soltysh) [inline pkg/controller/job/job_controller.go]: See my previous comment, this method is not needed.

**Review by soltysh — CHANGES_REQUESTED** [review]
I left you some comments.

File: pkg/controller/job/dynamicbackoff.go
Comment (clamoriniere1A) [reply pkg/controller/job/dynamicbackoff.go]: ok sure,  as discussed on slack. I will create two new methods in the current Backoff struct:
1) ```GetWithNbRetry()``` which returns backoff duration and number of retry.
2) ```NextWithInitDuration```which allow providing the default init backoff duration.

File: pkg/controller/job/job_controller_test.go
Comment (clamoriniere1A) [reply pkg/controller/job/job_controller_test.go]: yep

File: pkg/controller/job/job_controller_test.go
Comment (clamoriniere1A) [reply pkg/controller/job/job_controller_test.go]: block replaced by
```j.Spec.BackoffLimit = &backoffLimit```

File: pkg/controller/job/job_controller_test.go
Comment (clamoriniere1A) [reply pkg/controller/job/job_controller_test.go]: block replaced by
```j.Spec.BackoffSeconds = &backoffSeconds```

File: pkg/controller/job/job_controller.go
Comment (clamoriniere1A) [reply pkg/controller/job/job_controller.go]: method removed

File: pkg/controller/job/job_controller.go
Comment (clamoriniere1A) [reply pkg/controller/job/job_controller.go]: I created 2 new string vars for "reason" and "message" used in by ```Event()```and ```NewCondition()```

File: pkg/controller/job/job_controller.go
Comment (clamoriniere1A) [reply pkg/controller/job/job_controller.go]: done

File: pkg/controller/job/job_controller.go
Comment (clamoriniere1A) [reply pkg/controller/job/job_controller.go]: right

File: pkg/controller/job/job_controller.go
Comment (clamoriniere1A) [reply pkg/controller/job/job_controller.go]: right, thanks

**Comment by clamoriniere1A** [comment]
Hi @soltysh, I updated the PR with all your comments addressed.

File: staging/src/k8s.io/client-go/util/flowcontrol/backoff.go
Comment (soltysh) [inline staging/src/k8s.io/client-go/util/flowcontrol/backoff.go]: With this new method the other `Get` should be:
```go
func (p *Backoff) Get(id string) time.Duration {
    delay, _ := p.GetWithRetryNumber(id)
    return delay
}
```
We don't want copy & paste here.

File: staging/src/k8s.io/client-go/util/flowcontrol/backoff.go
Comment (soltysh) [inline staging/src/k8s.io/client-go/util/flowcontrol/backoff.go]: s/backoffDuration/initial

Let the name suggest what it's doing. Similarly to `NewBackOff`.

Comment (soltysh) [reply staging/src/k8s.io/client-go/util/flowcontrol/backoff.go]: Again, do not copy&paste code. Create a helper function and call that from `Next` and `NextWithInitDuration` passing `p.defaultDuration` or `initial` accordingly.

File: staging/src/k8s.io/client-go/util/flowcontrol/backoff.go
Comment (soltysh) [inline staging/src/k8s.io/client-go/util/flowcontrol/backoff.go]: `GetWithRetryNumber` seems better fitted name.

File: staging/src/k8s.io/client-go/util/flowcontrol/backoff.go
Comment (soltysh) [inline staging/src/k8s.io/client-go/util/flowcontrol/backoff.go]: @deads2k mind having a second look at these changes in BackOff?

File: staging/src/k8s.io/client-go/util/flowcontrol/backoff.go
Comment (soltysh) [inline staging/src/k8s.io/client-go/util/flowcontrol/backoff.go]: Not sure if this belongs here, but I'll defer to @deads2k for the final judgment.

**Review by soltysh — COMMENTED** [review]
Controller changes lgtm, I've asked David to review backoff changes.

File: staging/src/k8s.io/client-go/util/flowcontrol/backoff.go
Comment (clamoriniere1A) [reply staging/src/k8s.io/client-go/util/flowcontrol/backoff.go]: right, usually not me either.

File: staging/src/k8s.io/client-go/util/flowcontrol/backoff.go
Comment (clamoriniere1A) [reply staging/src/k8s.io/client-go/util/flowcontrol/backoff.go]: ok

File: staging/src/k8s.io/client-go/util/flowcontrol/backoff.go
Comment (deads2k) [inline staging/src/k8s.io/client-go/util/flowcontrol/backoff.go]: Can you pull this entire backoff.go change into a separate pull and get a review from the usual client suspects (chao, daniel, myself) with a good description of what its buying you. My initial impression is that is exposing internals that I wouldn't expect to see leaked.

File: pkg/controller/job/job_controller.go
Comment (deads2k) [inline pkg/controller/job/job_controller.go]: Using it for this?  I'd keep track of my retries locally to start.  Of a few dozen uses, you've found one that want this feature and its not obvious to me that all backoffs are created equal.  

Also, does `AddRateLimited` on the workqueue not do what you want?

File: staging/src/k8s.io/client-go/util/flowcontrol/backoff.go
Comment (clamoriniere1A) [reply staging/src/k8s.io/client-go/util/flowcontrol/backoff.go]: @deads2k thanks for the review, I created the separated pull request: https://github.com/kubernetes/kubernetes/pull/51626

File: pkg/controller/job/job_controller.go
Comment (clamoriniere1A) [reply pkg/controller/job/job_controller.go]: Looking at others controllers: ```federated.JobController``` and the ```Kubelet``` the usage of the ```flowcontrol.Backoff ``` was looking the best choice,

Adding a separated data-structure for storing the number of retries for each Job was a possibility that I envisaged. 
But the fact that we still need to update the ```Backoff``` structure in order to allow to set different initial delay durations for each entry, made me orient myself to integrate also the ```retry```counter in the ```Backoff``` struct

**Review by soltysh — APPROVED** [review]

**Comment by soltysh** [comment]
Rebased so 

/lgtm
/approve

**Comment by nikhiljindal** [comment]
The e2e test that was added in this PR ([sig-apps] Job should exceed backoffLimit) seems to be flaking recently: https://k8s-testgrid.appspot.com/release-1.8-blocking#gci-gce-1.8. 
@clamoriniere1A Can you please take a look?

**Comment by clamoriniere1A** [comment]
@nikhiljindal yes sure

**Pull request merged**
